File: day-4-scratchcards/part-2.ts
const NUMBER = /\d+/g;

function main(
	[_card, ...cards]: string[],
	[copy, ...copies]: string[],
): number {
	if (copy === undefined) return 0;

	const numbers = copy.split(':')[1].split('|');
	const first_numbers = numbers[0].match(NUMBER)!;
	const last_numbers = numbers[1].match(NUMBER)!;
	const intersection =
		first_numbers.filter((value) => last_numbers.includes(value)).length;

	const nested_count = main(cards, cards.slice(0, intersection));
	const sequential_count = main(cards, copies);

	return nested_count + sequential_count + 1;
}

const input = await Deno.readTextFile('input');
const cards = input.split('\n');

console.log(main(cards, cards));

File: day-4-scratchcards/part-2.hs
import Data.List
import Parsing

listOfNumbers = do
    number <- nat
    space
    numbers <- listOfNumbers
    return $ number : numbers
    <|> do
        number <- nat
        return [number]

gameParser = do
    string "Card"
    space
    id <- nat
    char ':'
    space
    firstNumbers <- listOfNumbers
    space
    char '|'
    space
    lastNumbers <- listOfNumbers
    return (id, firstNumbers, lastNumbers)

line :: [String] -> [String] -> IO Int
line _ [] = do return 0
line (_ : cards) (copy : copies) = do
    let (id, first_numbers, last_numbers) = fst $ head $ parse gameParser copy
    let intersection = length $ first_numbers `intersect` last_numbers

    nested_count <- line cards (take intersection cards) 
    sequencial_count <- line cards copies

    return $ nested_count + sequencial_count + 1

main = do
    input <- readFile "input"
    let cards = lines input

    sum <- line cards cards

    print sum
